segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,427 @@
+        Files.setPosixFilePermissions(Paths.get(config.getRoot(), handle.getSegmentName()), perms);
+        LoggerHelpers.traceLeave(log, "seal", traceId);
+        return null;
+    }
+
+    /**
+     * Concatenation as currently implemented here requires that we read the data and write it back to target file.
+     * We do not make the assumption that a native operation exists as this is not a common feature supported by file
+     * systems. As such, a concatenation induces an important network overhead as each byte concatenated must be
+     * read and written back when the storage is backed by a remote filesystem (through NFS).
+     *
+     * This option was preferred as other option (of having one file per transaction) will result in server side
+     * fragmentation and corresponding slowdown in cluster performance.
+     */
+    private Void syncConcat(SegmentHandle targetHandle, long offset, String sourceSegment) throws IOException {
+        long traceId = LoggerHelpers.traceEnter(log, "concat", targetHandle.getSegmentName(),
+                offset, sourceSegment);
+
+        Path sourcePath = Paths.get(config.getRoot(), sourceSegment);
+        Path targetPath = Paths.get(config.getRoot(), targetHandle.getSegmentName());
+
+        long length = Files.size(sourcePath);
+        try (FileChannel targetChannel = FileChannel.open(targetPath, StandardOpenOption.WRITE);
+             RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+            if (isWritableFile(sourcePath)) {
+                throw new IllegalStateException(String.format("Source segment (%s) is not sealed.", sourceSegment));
+            }
+            while (length > 0) {
+                final long retryOffset = offset;
+                final long retryLength = length;
+                long bytesTransferred = FileSystemRetryHelper.retry(
+                        () -> targetChannel.transferFrom(sourceFile.getChannel(), retryOffset, retryLength),
+                        (read) -> read == 0,
+                        () -> new IllegalStateException("Retries exceeded"),
+                        NUM_RETRIES);
+
+                offset += bytesTransferred;
+                length -= bytesTransferred;
+            }
+            Files.delete(sourcePath);
+            LoggerHelpers.traceLeave(log, "concat", traceId);
+            return null;
+        }
+    }
+
+    private Void syncDelete(SegmentHandle handle) throws IOException {
+        Files.delete(Paths.get(config.getRoot(), handle.getSegmentName()));
+        return null;
+    }
+
+    /**
+     * Executes the given supplier asynchronously and returns a Future that will be completed with the result.
+     */
+    private <R> CompletableFuture<R> supplyAsync(String segmentName, Callable<R> operation) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+
+        CompletableFuture<R> result = new CompletableFuture<>();
+        this.executor.execute(() -> {
```

Comment:
In the suggested case the `operation` needs to be `Supplier` which would reintroduce `SneakyThrow` in the `sync*` methods. As you mentioned both the operations are equivalent.